cli_tools/common/distro/distro.go

```diff
@@ -33,155 +36,216 @@ const (
-			l.minor == realOther.minor
-	}
-
-	if l.distro == sles {
-		return l.distro == realOther.distro &&
-			l.variant == realOther.variant &&
-			l.major == realOther.major
-	}
-
-	return l.distro == realOther.distro &&
-		l.major == realOther.major
+	// AsGcloudArg returns a hyphenated identifier, as used in the `--os` flag
+	// of `gcloud compute images import`. It is intended for
+	// showing to users in a help message. Notably, it lacks support for
+	// the -byol suffix, as that is not currently modeled.
+	AsGcloudArg() string
 }
 
-// FromLibguestfs initializes a DistroAndVersion from the fields returned by libguestfs's
-// inspection routines.
+// FromComponents interprets the (distro, major, minor) tuple and returns
+// a Release if the arguments are syntactically correct, and represent a
+// release that we *may* support. The caller is responsible for verifying
+// whether a translator is available elsewhere in the system.
 //
-// http://libguestfs.org/guestfs.3.html#guestfs_inspect_get_distro
-func FromLibguestfs(distro string, major string, minor string) (r Release, e error) {
+// Prefer this method for generating a new Release object. It is lenient, and
+// returns nice errors. The distro-specific constructors are strict and may panic.
+func FromComponents(distro string, major string, minor string) (r Release, e error) {
 	distro = strings.ToLower(distro)
-	if distro == windows {
+	majorInt, e := strconv.Atoi(major)
+	if e != nil || majorInt < 1 {
+		return nil, fmt.Errorf(
+			"major version required to be an integer greater than zero. Received: `%s`", major)
+	}
+	var minorInt int
+	if minor == "" {
+		minorInt = 0
+	} else {
+		minorInt, e = strconv.Atoi(minor)
+		if e != nil || minorInt < 0 {
+			return nil, errors.New(
+				"minor version required to be an integer greater than or equal to zero. Received: " + minor)
+		}
+	}
+	switch distro {
+	case "":
+		return nil, errors.New("distro name required")
+	case windows:
 		// We don't currently need Windows for this parsing, and punting
 		// since it's not immediately clear how to represent their
 		// version strings, as the user-facing value (Windows 2008r2)
 		// does not match what's used internally by our tools (NT 5.2)
 		// https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions
 		return r, errors.New("Windows not yet implemented")
+	case ubuntu:
+		return newUbuntuRelease(majorInt, minorInt)
+	case centos:
+		fallthrough
+	case debian:
+		fallthrough
+	case opensuse:
+		fallthrough
+	case rhel:
+		return newCommonLinuxRelease(distro, majorInt, minorInt)
 	}
-	if distro == "" || major == "" {
-		return nil, errors.New("distro and major are required")
+
+	// SLES supports variants, such as sles-sap.
+	if distro == sles || strings.HasPrefix(distro, sles+"-") {
+		return newSLESRelease(distro, majorInt, minorInt)
 	}
-	return release{
-		distro: distro,
-		major:  major,
-		minor:  minor,
-	}, nil
+
+	return nil, fmt.Errorf("distro `%s` is not importable", distro)
```

Comment:
Is `distro` a common term used by customers or would `distribution` be more appropriate?

Reply:
Similar to the question above, this isn't shown to users at the moment, and it isn't written with that in mind.